File: python/productExceptSelf.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
      numsLen = len(nums)
      l, r = [0] * numsLen, [0] * numsLen

      # set up left half
      l[0] = 1
      for i in range(1, numsLen):
        l[i] = l[i-1] * nums[i-1]

      # set up right half
      r[numsLen-1] = 1
      for j in reversed(range(numsLen - 1)):
        r[j] = r[j+1] * nums[j+1]

      # combine the products for both the left and right half
      for k in range(numsLen):
        nums[k] = l[k] * r[k]

      return nums

File: go/productExceptSelf.go
func productExceptSelf(nums []int) []int {
	// the premise of the problem is to calculate left and right subarrays
	left := make([]int, len(nums))
	right := make([]int, len(nums))

	// find all the products left of array
	left[0] = 1
	for i := 1; i < len(nums); i++ {
		left[i] = left[i-1] * nums[i-1]
	}

	// find all products right of the array
	right[len(nums)-1] = 1
	for j := len(nums) - 2; j >= 0; j-- {
		right[j] = right[j+1] * nums[j+1]
	}

	result := make([]int, len(nums))
	for i := range result {
		result[i] = left[i] * right[i]
	}

	return result
}
